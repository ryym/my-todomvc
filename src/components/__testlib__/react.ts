import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

// Though we used @testing-library/react first, it looks like it has a bug
// about the change event simulation (https://github.com/testing-library/react-testing-library/issues/359).
// And the official react-dom/test-utils is useful enough.

export const render = (element: ReactElement<any>) => {
  const container = document.createElement('div');
  act(() => {
    ReactDOM.render(element, container);
  });
  return container;
};

export const getByTestId = (container: HTMLElement, testId: string): HTMLElement => {
  const element = container.querySelector(`[data-testid="${testId}"]`) as HTMLElement;
  if (element == null) {
    throw new Error(`Test ID not found: ${testId}`);
  }
  return element;
};

export const renderSnapshot = (element: ReactElement<any>) => {
  return renderer.create(element).toJSON();
};
