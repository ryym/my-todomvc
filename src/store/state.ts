import { Todos } from '../lib/Todos';
import { TodoState } from '../lib/TodoState';

export interface State {
  readonly todos: TodosState;
  readonly filters: FiltersState;
}

export type TodosState = Todos;

export interface FiltersState {
  readonly todoState: TodoState.Value;
}

export const initFilters = (): FiltersState => ({
  todoState: TodoState.Value.All,
});
