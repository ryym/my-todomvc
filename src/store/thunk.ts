import { Thunk as RedyThunk } from 'redy';
import { State } from './state';

export type Thunk<R = void> = RedyThunk<State, R>;
