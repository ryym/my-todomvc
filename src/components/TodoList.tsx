import React, { useState } from 'react';
import { connect } from '../store/connect';
import { selectFilteredTodos } from '../store/selectors/todos';
import { TodoItem, EditState } from './TodoItem';

export const TodoList = connect(
  state => ({
    todos: selectFilteredTodos(state),
  }),

  function TodoList({ dispatch, todos }) {
    const [editedTodoId, setEditedTodoId] = useState<number | null>(null);

    const handleEditStateChange = (state: EditState) => {
      switch (state.type) {
        case 'EditStart':
          setEditedTodoId(state.id);
          break;
        case 'EditEnd':
          setEditedTodoId(null);
          break;
      }
    };

    return (
      <ul className="todo-list">
        {todos.map(todo => (
          <TodoItem
            key={todo.id}
            todo={todo}
            editable={editedTodoId === todo.id}
            onEditStateChange={handleEditStateChange}
            dispatch={dispatch}
          />
        ))}
      </ul>
    );
  }
);
