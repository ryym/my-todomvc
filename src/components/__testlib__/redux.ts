import { Dispatch, Store, createStore } from 'redux';
import { State } from '../../store/state';

export function mockStore<S = State>(initialState: S, dispatch?: Dispatch): Store<S> {
  const noopReducer = (state: any) => state;
  let mockStore = createStore(noopReducer, initialState);
  if (dispatch) {
    mockStore = { ...mockStore, dispatch };
  }
  return mockStore;
}
