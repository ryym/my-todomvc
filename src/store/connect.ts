import { SFC } from 'react';
import { Dispatch } from 'redux';
import { connect as reduxConnect } from 'react-redux';
import { State } from './state';

export type WithDispatch<P extends {}> = P & { dispatch: Dispatch };

export function connect<WrapperProps extends {}, Props extends {} = {}>(
  mapStateToProps: (state: State, props: WrapperProps) => Props,
  wrappedComponent: SFC<WithDispatch<Props>>
): SFC<WrapperProps> {
  return (reduxConnect as any)(mapStateToProps)(wrappedComponent);
}
