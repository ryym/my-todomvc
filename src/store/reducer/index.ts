import { combineReducers, Reducer } from 'redux';
import { reduceTodos } from './todos';
import { reduceFilters } from './filters';
import { State } from '../state';

export const createReducer = (): Reducer<State> => {
  return combineReducers({
    todos: reduceTodos,
    filters: reduceFilters,
  });
};
