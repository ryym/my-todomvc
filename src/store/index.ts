import { createStore, applyMiddleware, compose, Store } from 'redux';
import { redyMiddleware } from 'redy';
import logger from 'redux-logger';
import { createReducer } from './reducer';
import { State } from './state';

// https://github.com/zalmoxisus/redux-devtools-extension#installation
const composeEnhancers: typeof compose =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const configureStore = (initialState: Partial<State>): Store<State> => {
  const reducer = createReducer();
  const enhancer = composeEnhancers(applyMiddleware(logger, redyMiddleware()));
  return createStore(reducer, initialState, enhancer);
};
