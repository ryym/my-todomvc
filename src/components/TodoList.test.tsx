import React from 'react';
import { Provider } from 'react-redux';
import { TodoList } from './TodoList';
import { initState } from '../store/state';
import { mockStore } from './__testlib__/redux';
import { renderSnapshot } from './__testlib__/react';

it('matches snapshot', () => {
  const store = mockStore({
    ...initState(),
    todos: [{ id: 1, title: 'hello', completed: false }],
  });
  const tree = renderSnapshot(
    <Provider store={store}>
      <TodoList />
    </Provider>
  );
  expect(tree).toMatchSnapshot();
});
