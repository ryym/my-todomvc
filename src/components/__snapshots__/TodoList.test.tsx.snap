// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<ul
  className="todo-list"
>
  <li
    className=""
  >
    <div
      className="view"
    >
      <input
        checked={false}
        className="toggle"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        data-testid="title-label"
        onDoubleClick={[Function]}
      >
        hello
      </label>
      <button
        className="destroy"
        onClick={[Function]}
      />
    </div>
    <input
      className="edit"
      data-testid="title-edit"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value="hello"
    />
  </li>
</ul>
`;
